src/v/datalake/translation/state_machine.cc

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/translation/state_machine.h"
+
+#include "datalake/logger.h"
+#include "datalake/translation/types.h"
+
+namespace {
+raft::replicate_options make_replicate_options() {
+    auto opts = raft::replicate_options(raft::consistency_level::quorum_ack);
+    opts.set_force_flush();
+    return opts;
+}
+
+model::record_batch_reader make_translation_state_batch(kafka::offset offset) {
+    auto val = datalake::translation::translation_state{
+      .highest_translated_offset = offset};
+    storage::record_batch_builder builder(
+      model::record_batch_type::datalake_translation_state, model::offset(0));
+    builder.add_raw_kv(std::nullopt, serde::to_iobuf(val));
+    auto batch = std::move(builder).build();
+    return model::make_memory_record_batch_reader(std::move(batch));
+}
+
+} // namespace
+
+namespace datalake::translation {
+
+translation_stm::translation_stm(ss::logger& logger, raft::consensus* raft)
+  : raft::persisted_stm<>("datalake_translation_stm.snapshot", logger, raft) {}
+
+ss::future<> translation_stm::do_apply(const model::record_batch& batch) {
+    if (
+      batch.header().type
+      != model::record_batch_type::datalake_translation_state) {
+        co_return;
+    }
+    auto records = batch.copy_records();
+    vassert(
+      records.size() == 1,
+      "Invalid metadata in batch {}, size: {}",
+      batch.header(),
+      records.size());
+    auto& r = *records.begin();
+    auto val = serde::from_iobuf<translation_state>(r.release_value());
+    vlog(
+      datalake_log.trace,
+      "updating highest translated offset to {}",
+      val.highest_translated_offset);
+    _highest_translated_offset = val.highest_translated_offset;
```

Comment:
Should we std::max these? If other code remains bug free forever I imagine it makes no difference, but that's a big if